File: src/commands/push.rs
use crate::config::Config;
use crate::utils::{get_git_root, count_unpushed_commits};
use anyhow::{Context, Result};
use git2::Repository;
use std::path::Path;

/// 推送模块仓库变更
pub fn execute(module: Option<String>) -> Result<()> {
    let root = get_git_root()?;
    let config_path = Path::new(&root).join(".gitmrepo");

    // 检查配置文件是否存在
    if !config_path.exists() {
        println!("⚠️  .gitmrepo 配置文件不存在");
        println!("   请先执行: git mrepo init");
        return Ok(());
    }

    // 加载配置
    let config = Config::load(&config_path)?;

    // 确定要操作的模块
    let modules = if let Some(name_or_path) = module {
        vec![config.find_module(&name_or_path)?.clone()]
    } else {
        if config.modules.is_empty() {
            println!("⚠️  未注册任何模块仓库");
            return Ok(());
        }
        config.modules.values().cloned().collect()
    };

    println!("🔄 推送模块仓库变更...");

    for module in modules {
        println!("\n[{}] {}:", module.name, module.path);

        // 检查目录是否存在
        if !Path::new(&module.path).exists() {
            println!("  ⚠️  目录不存在");
            continue;
        }

        // 检查未推送提交
        let repo = Repository::open(&module.path)
            .context("无法打开 Git 仓库")?;

        let unpushed_count = count_unpushed_commits(&repo, &module.branch)?;

        if unpushed_count == 0 {
            println!("  ✅ 没有需要推送的提交");
            continue;
        }

        println!("  未推送提交: {}", unpushed_count);

        // 执行 git push
        let output = std::process::Command::new("git")
            .current_dir(&module.path)
            .args(&["push", "origin", &module.branch])
            .output()
            .context("执行 git push 失败")?;

        if output.status.success() {
            println!("  ✅ 已推送 {} 个提交", unpushed_count);
        } else {
            let stderr = String::from_utf8_lossy(&output.stderr);
            println!("  ❌ {}", stderr);
        }
    }

    Ok(())
}

File: src/commands/push.ts
import * as path from 'node:path';
import * as fs from 'node:fs';
import { execSync } from 'node:child_process';
import { ConfigManager, type Module } from '../config.js';
import { getGitRoot, countUnpushedCommits } from '../utils/index.js';

/**
 * 执行 push 命令 - 推送模块仓库变更
 */
export async function pushExecute(moduleArg?: string): Promise<void> {
  const root = getGitRoot();
  const configPath = path.join(root, '.gitmrepo');

  if (!fs.existsSync(configPath)) {
    console.log('⚠️  .gitmrepo 配置文件不存在');
    console.log('   请先执行: git mrepo init');
    return;
  }

  const config = ConfigManager.load(configPath);
  const cm = new ConfigManager();

  // 确定模块列表
  const modules: Module[] = moduleArg
    ? [(() => {
        const m = cm.findModule(config, moduleArg);
        if (!m) throw new Error(`模块不存在: ${moduleArg}`);
        return m;
      })()]
    : Object.values(config.modules);

  if (modules.length === 0) {
    console.log('⚠️  没有模块需要推送');
    return;
  }

  console.log('🔄 推送模块仓库变更...');

  for (const module of modules) {
    console.log(`[${module.name}] ${module.path}:`);

    const fullPath = path.isAbsolute(module.path) ? module.path : path.join(root, module.path);

    if (!fs.existsSync(fullPath)) {
      console.log('  ⚠️  目录不存在');
      continue;
    }

    if (!fs.existsSync(path.join(fullPath, '.git'))) {
      console.log('  ⚠️  不是 Git 仓库（缺少 .git），跳过');
      continue;
    }

    const unpushedCount = countUnpushedCommits(fullPath, module.branch);

    if (unpushedCount === 0) {
      console.log('  ✅ 没有需要推送的提交');
    } else {
      console.log(`  未推送提交: ${unpushedCount}`);
    }

    try {
      execSync(`git push origin "${module.branch}"`, {
        cwd: fullPath,
        encoding: 'utf-8',
      });
      console.log(`  ✅ 已推送 ${unpushedCount} 个提交`);
    } catch (error) {
      const stderr = (error as any).stderr?.toString() || '未知错误';
      console.log(`  ❌ ${stderr}`);
    }
  }
}
